### PR TITLE
Added support to handle Temple/Portal in the editor + deharcoded workers.

### PR DESCRIPTION
**Pull request by Bertram25**
Part of #192

I also added support to handle moving the portal and temple only within the editor.
But this still triggers an assert failure when removing tiles including the one
with the mesh.
E.g.:
ERROR: Assert failed file opendungeons/source/network/ODClient.cpp line 622 info : name=RenderedMovableEntity_Portal2_50

I've made a few tests but so far I really can't tell what's triggering it.
Any advice? ;)

Otherwise, it' working fine.

### CODE REVIEW

**Comment by hwoarangmy** [comment]
Concerning the assert/segfault problem, if it only happens when removing tiles, I would guess that you are removing the dungeon temple/portal object twice and that's why it is not found the second time. Could you please post some log ?
If you want, I can have a look tonight.

**Comment by hwoarangmy** [comment]
Since the FOW, as you have seen in the PR, the objects creation/remove has become a bit tricky. So, there might be some bug hidding. Moreover, I've not tried to manually remove objects like it is done here. That's why we can also imagine that there is some bug or incompatibility with the architecture.

**Comment by Bertram25** [comment]
> I would guess that you are removing the dungeon temple/portal object twice and that's why it is not found the second time. Could you please post some log ?

Ok, I'll do it, I just wanted to know whether I did something plainly stupid. I do think it is a good way to practice with objects with the latest architecture, though. :)

> That's why we can also imagine that there is some bug or incompatibility with the architecture.

That's the thing I feared. Doing something incompatible with the current architecture. Also, here, I first wanted to only move the temple/portal and I changed the update*Position() accordingly to use a setPosition(there, isMoving = true); on the temple object but this is not working. That's why I changed the logic to delete/recreate the object each time.
The second problem I had is that CreateMeshLocal() is called each time before absorbRoom() if there is an absorption, so I had to let the mesh creation there, while I obtained a duplicate when the room is absorbed.
Maybe you have a better way to do it? (As you basically rewrote that system, eh. ;])

Best regards,

**Comment by hwoarangmy** [comment]
Well, Building::updateActiveSpots was supposed to handle active spots. It is overriden in Room and it is the function that removes/creates the activespots when needed (by calling notifyActiveSpotCreated and notifyActiveSpotRemoved). Since the active spots behaviour for Portal/Dungeon temple is pretty specific, IMHO, the easiest would be to override Room::updateActiveSpots (which I forgot to set as virtual btw). IIRC, it is called only after room absorbtion and it works with the editor so it is probably well fitted for that ^^

**Comment by hwoarangmy** [comment]
While I'm thinking about it, if you override Room::updateActiveSpots, you should remember that you will have extra calls when a neighboor wall is claimed. That means that you would need to check if the room object has moved before removing/adding it

**Comment by Bertram25** [comment]
Thanks for the hints, I'll redo the patch. :)

**Comment by Bertram25** [comment]
Should be working good now. :)

**Comment by akien-mga** [comment]
It works fine here. Room merging seems to work great with all kind of room sizes, even though portals and temples are meant to be on 3x3 rooms.

**Comment by Bertram25** [comment]
> It works fine here. Room merging seems to work great with all kind of room sizes, even though portals and temples are meant to be on 3x3 rooms.

Cool :)

> > Ideally, the mode ougt to be passed in as a parameter where it's needed.
> > That's what I've done when I had to but I have to say that updateActivespots is used in every trap/room so that will mean many changes...

So yes, I'd rather add a function in the gamemap class, since it's the true class making the most the separation between the two threads logic. What about:

``` c++
bool GameMap::isInEditorMode() const
{
    if (isServerGameMap())
    {
        return (ODServer::getServerMode == ServerMode::ModeEditor)
    }

    return (ODFrameListener::getSingleton().getModeManager->getCurrentModeTypeExceptConsole() == Modetype::EDITOR)
}
```

**Comment by hwoarangmy** [comment]
> I'd rather add a function in the gamemap class, since it's the true class making the most the separation between the two threads logic

That seems ok to me

**Comment by Bertram25** [comment]
Note to self:
- Add function to test whether one is in map editing mode in the game map object.
- Add support to define a default worker class in Seat, when initialization the spawn pool.

I'll do that asap. :)

**Comment by hwoarangmy** [comment]
If I am not mistaken, you have done all you wanted in this PR, right ?
If it is the case, I will test tonight ^^

**Comment by Bertram25** [comment]
Yep, normally I'm done. :)

**Comment by hwoarangmy** [comment]
Seems like rebasing is needed ^^

**Comment by Bertram25** [comment]
k, I do that asap. :)

**Comment by Bertram25** [comment]
Rebased :)

**Comment by hwoarangmy** [comment]
Nice. I will test tonight

File: source/utils/ConfigManager.h [hunk 1]
```diff
@@ -130,6 +134,10 @@ class ConfigManager : public Ogre::Singleton<ConfigManager>
     std::map<const CreatureDefinition*, std::vector<const SpawnCondition*> > mCreatureSpawnConditions;
     std::map<const std::string, std::vector<const CreatureMood*> > mCreatureMoodModifiers;
     std::map<const std::string, std::vector<std::string> > mFactionSpawnPool;
+
+    //! \brief Stores the faction default worker creature definition.
+    std::map<const std::string, std::string> mFactionDefaultWorkerClass;
+
```
Comment (oyvindln) [inline source/utils/ConfigManager.h:140]: Not that it's going to make a huge difference, but the strings in the map probably shouldn't be const, as that may prevent move-construction if the map has to reallocate memory.

Comment (hwoarangmy) [inline source/utils/ConfigManager.h:140]: For maps? I thought it was for vectors only...

Comment (oyvindln) [inline source/utils/ConfigManager.h:140]: Hm, you are right about maps it seems.

**Comment by hwoarangmy** [comment]
Works well, I merge

**Pull request merged**